### PR TITLE
#5044: Add optional output tensor for backward ops

### PR DESCRIPTION
**Pull request by bharane-ab**
Add  optional output tensor support for backward op addcmul.
Primary issue : #5044

### CODE REVIEW

File: tt_eager/tt_dnn/op_library/backward/backward_ops.cpp [hunk 1]
```diff
@@ -119,21 +119,41 @@ std::vector<Tensor> exp_bw(const Tensor& grad, const Tensor& input, const Memory
...
+std::vector<std::optional<Tensor>> _addcmul_bw(const Tensor& grad, const Tensor& input, const Tensor& tensor1, const Tensor& tensor2, float value, const MemoryConfig& output_mem_config, const std::vector<bool> are_required_outputs, std::vector<std::optional<Tensor>> output_tensor) {
+    if(are_required_outputs.at(0)){
+        output_tensor.at(0) = grad;
+    }
+    else
+    {
+        output_tensor.at(0) = std::nullopt;
+    }
+    if(are_required_outputs.at(1)){
+        output_tensor.at(1) = mul_unary(mul(grad, tensor2, std::nullopt, output_mem_config), value, output_mem_config);
```
Comment (eyonland) [inline tt_eager/tt_dnn/op_library/backward/backward_ops.cpp:131]: output_tensor.at(1) = mul(grad, tensor2, std::nullopt, output_mem_config, std::move(output_tensor.at(1)));
ouptut_tensor.at(1) = mul_unary(output_tensor.at(1), value, ouptut_mem_config, output_tensor.at(1));

Comment (eyonland) [reply tt_eager/tt_dnn/op_library/backward/backward_ops.cpp:131]: If you use other ops then we need to reuse the buffers preallocated for the optional output tensors.

File: tt_eager/tt_dnn/op_library/backward/backward_ops.cpp [hunk 2]
```diff
@@ -119,21 +119,43 @@ std::vector<Tensor> exp_bw(const Tensor& grad, const Tensor& input, const Memory
...
 
 
-std::vector<Tensor> _addcmul_bw(const Tensor& grad, const Tensor& input, const Tensor& tensor1, const Tensor& tensor2, float value, const MemoryConfig& output_mem_config) {
-    std::vector<Tensor> grad_tensor;
-    grad_tensor.emplace_back(grad);
-    Tensor grad_a = mul_unary(mul(grad, tensor2, std::nullopt, output_mem_config), value, output_mem_config);
-    grad_tensor.emplace_back(grad_a);
-    Tensor grad_b = mul_unary(mul(grad, tensor1, std::nullopt, output_mem_config), value, output_mem_config);
-    grad_tensor.emplace_back(grad_b);
+std::vector<std::optional<Tensor>> _addcmul_bw(const Tensor& grad, const Tensor& input, const Tensor& tensor1, const Tensor& tensor2, float value, const MemoryConfig& output_mem_config, const std::vector<bool> are_required_outputs, std::vector<std::optional<Tensor>> output_tensor) {
```
Comment (eyonland) [inline tt_eager/tt_dnn/op_library/backward/backward_ops.cpp:122]: Is it possible to return a `std::optional<Tensor>` and not return a std::vector of these?

**Review by eyonland — CHANGES_REQUESTED** [review]
Is it possible to return a `std::optional<Tensor>` and not return a std::vector of these?

File: tests/tt_eager/python_api_testing/unit_testing/backward_ops/test_backward_addcdiv.py [hunk 3]
```diff
@@ -40,3 +40,54 @@ def test_bw_addcdiv(input_shapes, value, device):
...
+@pytest.mark.parametrize(
+    "input_shapes",
+    (
+        (torch.Size([1, 1, 32, 32])),
+        (torch.Size([1, 1, 320, 384])),
+        (torch.Size([1, 3, 320, 384])),
+    ),
+)
+@pytest.mark.parametrize("value", [0.12])
+@pytest.mark.parametrize("are_required_outputs", [[True, True, True], [False, True, True], [False, False, False]])
```
Comment (seunghwan100) [inline tests/tt_eager/python_api_testing/unit_testing/backward_ops/test_backward_addcdiv.py:54]: Why don't you test all combinations?

File: tests/tt_eager/python_api_testing/unit_testing/backward_ops/test_backward_addcmul.py [hunk 4]
```diff
@@ -39,3 +39,54 @@ def test_bw_addcmul(input_shapes, value, device):
...
+@pytest.mark.parametrize(
+    "input_shapes",
+    (
+        (torch.Size([1, 1, 32, 32])),
+        (torch.Size([1, 1, 320, 384])),
+        (torch.Size([1, 3, 320, 384])),
+    ),
+)
+@pytest.mark.parametrize("value", [0.05, 1.0, 0.5, 0.12])
+@pytest.mark.parametrize("are_required_outputs", [[True, True, True], [False, True, True], [False, False, False]])
```
Comment (seunghwan100) [inline tests/tt_eager/python_api_testing/unit_testing/backward_ops/test_backward_addcmul.py:53]: Why don't you test all combinations?

File: tt_eager/tt_dnn/op_library/backward/backward_ops.cpp [hunk 5]
```diff
@@ -119,21 +119,43 @@ std::vector<Tensor> exp_bw(const Tensor& grad, const Tensor& input, const Memory
...
+    else
+    {
+        output_tensor.at(2) = std::nullopt;
+    }
 
-    return grad_tensor;
+    return output_tensor;
 }
-std::vector<Tensor> addcmul_bw(const Tensor& grad, const Tensor& input, const Tensor& tensor1, const Tensor& tensor2, float value, const MemoryConfig& output_mem_config)
+std::vector<std::optional<Tensor>> addcmul_bw(const Tensor& grad, const Tensor& input, const Tensor& tensor1, const Tensor& tensor2, float value, const MemoryConfig& output_mem_config, const std::vector<bool> are_required_outputs, std::vector<std::optional<Tensor>> output_tensor)
```
Comment (seunghwan100) [inline tt_eager/tt_dnn/op_library/backward/backward_ops.cpp:149]: I think this is better.
```const std::vector<bool>& are_required_outputs```

File: tt_eager/tt_dnn/op_library/backward/backward_ops.cpp [hunk 6]
```diff
@@ -119,21 +119,43 @@ std::vector<Tensor> exp_bw(const Tensor& grad, const Tensor& input, const Memory
...
+std::vector<std::optional<Tensor>> _addcmul_bw(const Tensor& grad, const Tensor& input, const Tensor& tensor1, const Tensor& tensor2, float value, const MemoryConfig& output_mem_config, const std::vector<bool> are_required_outputs, std::vector<std::optional<Tensor>> output_tensor) {
+    if(are_required_outputs.at(0)){
+        output_tensor.at(0) = grad;
+    }
+    else
+    {
+        output_tensor.at(0) = std::nullopt;
+    }
+    if(are_required_outputs.at(1)){
+        output_tensor.at(1) = mul(grad, tensor2, std::nullopt, output_mem_config);
```
Comment (dongjin-na) [inline tt_eager/tt_dnn/op_library/backward/backward_ops.cpp:131]: Please refer to this [comment](https://github.com/tenstorrent/tt-metal/pull/8671/files#r1615205319) for more details.

File: tt_eager/tt_dnn/op_library/backward/backward_ops.cpp (hunk 2 above)
Comment (bharane-ab) [reply tt_eager/tt_dnn/op_library/backward/backward_ops.cpp]: As backward ops send multiple output like grad, input_a and input_b we need the return type as vector.

**Comment by ruthreshx** [comment]
@eyonland and @KalaivaniMCW 
Reg the addcmul and addcdiv was redirected to [PR](https://github.com/tenstorrent/tt-metal/pull/9600)

**Review by eyonland — APPROVED** [review]

**Pull request closed without merging**